lib/Parse/Lexer.cpp

```diff
@@ -819,10 +819,17 @@ void Lexer::lexDollarIdent() {
     }
   }
 
-  // It's always an error to see a standalone $
   if (CurPtr == tokStart + 1) {
-    diagnose(tokStart, diag::expected_dollar_numeric);
-    return formToken(tok::unknown, tokStart);
+    // It is never an error to see a standalone '$' when in Swift 3
+    // compatibility mode.
+    if (LangOpts.isSwiftVersion3()) {
+      return formToken(tok::identifier, tokStart);
+    } else {
+      // Else offer to replace '$' with '`$`'.
+      diagnose(tokStart, diag::standalone_dollar_identifier)
+        .fixItReplaceChars(getSourceLoc(tokStart), getSourceLoc(CurPtr), "`$`");
+      return formToken(tok::unknown, tokStart);
```

Comment:
I was hoping we could find a way to suppress the double diagnostic. If we return `tok::unknown`, that code is basically a parsing train-wreck, and we get no more useful diagnostics out of it. If we return `tok::identifier`, everything else works fine except that there are a few Fix-Its to apply. The latter is a much better experience, particularly in Xcode where one can "Fix All In Scope".

I think the lexer has a general problem with diagnostics and backtracking. It wouldn't be totally unreasonable to hold on to a set of locations at which diagnostics were emitting, and check that right before emitting the diagnostics.